Implement 'Nil User' pattern to deal with deleted users

radar opened the pull request:
Potential fix for #427.

I think this is a good way to deal with deleted users. @knewter, @parndt, OK?

radar committed:
Implement 'Nil User' pattern to deal with deleted users

Potential fix for #427

parndt commented:
In general seems good. 

On Tuesday, 28 May at 3:19 PM Ryan Bigg wrote:

> Potential fix for #427.
> 
> I think this is a good way to deal with deleted users. @knewter, @parndt, OK?
> You can merge this Pull Request by running:
> 
>   git pull https://github.com/radar/forem niluser-pattern
> 
> Or you can view, comment on it, or merge it online at:
> 
>   https://github.com/radar/forem/pull/431
> 
> -- Commit Summary --
> - Implement &#39;Nil User&#39; pattern to deal with deleted users
> 
> -- File Changes --
> 
> ```
> A app/models/forem/concerns/nil_user.rb (13)
> A app/models/forem/nil_user.rb (11)
> M app/models/forem/post.rb (3)
> M app/models/forem/topic.rb (3)
> M spec/requests/posts_spec.rb (11)
> M spec/requests/topic_listing_spec.rb (20)
> ```
> 
> -- Patch Links --
> 
> https://github.com/radar/forem/pull/431.patch
> https://github.com/radar/forem/pull/431.diff

radar committed:
Use Workflow 1.0.0 and change update_attribute calls to update_column calls.

There's also been a change to use Workflow's hook methods, so that Post and Topic both have their own #approve methods now.

radar commented:
I've snuck in another change here which was required to fix some of the tests for the nil user.

It updates Forem to use Workflow 1.0.0 and switches some update_attribute calls to update_column calls so that the callbacks aren't triggered _all the damn time_. Workflow also provides the ability to define event methods which allows us to avoid using typical ActiveRecord callbacks for our state machine events.

radar commented on spec/support/factories/topics.rb:
```diff
@@ -5,10 +5,10 @@
-      state 'approved'
+    factory :approved_topic do
+      after_create do |t|
+        t.approve!
```
This is necessary to trigger the `approve` method callbacks.

parndt commented on app/views/forem/posts/_post.html.erb:
```diff
@@ -3,7 +3,11 @@
+        <%= t(:deleted) %>
+      <% else %>
+        <%= link_to_if Forem.user_profile_links, post.user, [main_app, post.user] %>
+      <% end %>
```
Normally a NilUser would make the view less complicated ;-)

avitus commented:
Would be great if this could be merged.

parndt commented:
My only gripe was the view complication.

radar commented:
Added to master and rails4. If anyone has any ideas on how to make the view less complicated, patches are welcome :)
